posts/2022-07-25-keyword-generics.md

```diff
@@ -0,0 +1,344 @@
+_bready_, but that would make for a weird-looking sandwich. Ergo: sandwich
+dilemma. (yes, you can ignore all of this.)
+
+```rust
+enum Option<T> {
+    Some(T),
+    None,
+}
+
+impl<T> Option<T> {
+    fn map<J>(self, f: impl FnOnce(T) -> J) -> Option<J> { ... }
+}
+```
+
+```rust
+my_option.map(|x| x.await)
+```
+
+This will produce a compiler error: the closure `f` is not an async context, so
+`.await` cannot be used within it. And we can't just convert the closure to be
+`async` either, since `fn map` doesn't know how to call async functions. In
+order to solve this issue, we could provide a new `async_map` method which
+_does_ provide an async closure. But we may want to repeat those for more
+effects, and that would result in a combinatorial explosion of effects. Take for
+example "can fail" and "can be async":
+
+|                | not async    | async              |
+| -------------- | ------------ | ------------------ |
+| __infallible__ | `fn map`     | `fn async_map`     |
+| __fallible__   | `fn try_map` | `fn async_try_map` |
+
+That's a lot of API surface for just a single method, and __that problem
+multiplies across the entire API surface in the stdlib__. We expect that once we
+start applying "keyword generics" to traits, we will be able to solve the
+sandwich problem. The type `f` would be marked generic over a set of effects,
+and the compiler would choose the right variant during compilation.
+
+## Affecting all effects
+
+Both `const` and `async` share a very similar issue, and we expect that other
+"effects" will face the same issue. "fallibility" particularly on our mind here,
+but it isn't the only effect. In order for the language to feel consistent we
+need consistent solutions.
+
+## FAQ
+
+### Q: Will this make the language more complicated?
+
+The goal of keyword generics is not to minimize the complexity of the Rust
+programming language, but to _minimize the complexity of programming in Rust._
+These two might sound similar, but they're not. Our reasoning here is that by
+_adding_ a feature, we will actually be able to significantly reduce the surface
+area of the stdlib, crates.io libraries, and user code - leading to a more
+streamlined user experience.
+
+Choosing between sync or async code is a fundamental choice which needs to be
+made. This is complexity which cannot be avoided, and which needs to exist
+somewhere. Currently in Rust that complexity is thrust entirely on users of
+Rust, making them responsible for choosing whether their code should support
+async Rust or not. But other languages have made diferent choices. For example
+Go doesn't distinguish between "sync" and "async" code, and has a runtime which
+is able to remove that distinction.
+
+The work we're doing would make it so that complexity of choosing between sync
+and async can be moved from user code into the type system, and from there on
+out can be handled by the compiler instead.
```

Comment:
I feel like this isn't quite right -- presumably, the user still needs to choose what kind of I/O model they're using, since that has major performance (at least) implications, and plausibly correctness ones around timeouts etc.

I think the right claim earlier would be that the "end/final/binary" user must choose, but perhaps the individual developer of a library (e.g., of serde) may not need to choose. That resonates more with me.

Reply:
> I think the right claim earlier would be that the "end/final/binary" user must choose, but perhaps the individual developer of a library (e.g., of serde) may not need to choose. That resonates more with me.

Yeah, that's exactly what we're going for as well. I agree that the current phrasing doesn't quite capture that. Let me find a way to update that.